### PR TITLE
Allow negative axes in reductions module

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

## Category:

**New feature**

## Description:

- Adds support for negative indices in the `axes` argument of reduction operators
- Accepted range is [-ndim, ndim-1]
- Negative indices are counted from the back.

### Affected modules and functionalities:

reductions.* ops

### Key points relevant for the review:

Everything

### Tests:

test_reduce.test_reduce
- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [x] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: REDUCE.05

**JIRA TASK**: DALI-3139

### CODE REVIEW

**Review by awolant — APPROVED** [review]

File: dali/kernels/reduce/reduce_cpu.h [hunk 1]
```diff
@@ -236,13 +237,15 @@ struct ReduceBaseCPU {
 
   void InitAxes(span<const int> _axes) {
     for (int axis : _axes) {
-      if (axis < 0 || axis >= ndim()) {
-        throw std::range_error(make_string("Axis index out of range: ", axis, " not in 0..",
-                               ndim()-1));
+      if (axis < -ndim() || axis >= ndim()) {
+        throw std::range_error(
+            make_string("Axis index out of range: ", axis, " not in ", -ndim(), "..", ndim() - 1));
```
Comment (szalpal) [inline dali/kernels/reduce/reduce_cpu.h:242]: How about applying mathematical range notation?
```suggestion
            make_string("Axis index out of range: ", axis, " not in [", -ndim(), ", ", ndim() - 1, "]"));
```

File: dali/kernels/reduce/reduce_setup_utils.h [hunk 2]
```diff
@@ -148,15 +148,33 @@ inline void CheckAxes(span<const int> axes, int ndim) {
   assert(ndim >= 0 && ndim <= 64);
   uint64_t mask = 0;
   for (auto a : axes) {
-    if (a < 0 || a >= ndim)
-      throw std::out_of_range(make_string("Axis index out of range: ", a, " not in 0..", ndim-1));
+    if (a < -ndim || a >= ndim)
+      throw std::out_of_range(make_string("Axis index out of range: ", a, " not in ", -ndim, "..", ndim-1));
```
Comment (szalpal) [inline dali/kernels/reduce/reduce_setup_utils.h:152]: How about applying mathematical range notation?
```suggestion
      throw std::out_of_range(make_string("Axis index out of range: ", axis, " not in [", -ndim(), ", ", ndim() - 1, "]"));
```

File: dali/test/python/operator/test_reduce.py [hunk 3]
```diff
@@ -226,6 +230,14 @@ def test_reduce():
                 for type_id in types:
                     yield run_reduce, keep_dims, reduction_name, batch_gen, type_id
 
+def test_reduce_negative_axes():
+    reductions = ["sum", "max"]
+    type = np.uint8
```
Comment (szalpal) [inline dali/test/python/operator/test_reduce.py:235]: How about adding a test for improper axes? Like `[-100]` or something along?

**Review by szalpal — APPROVED** [review]

File: dali/kernels/reduce/reduce_setup_utils.h (hunk 2 above)
Comment (jantonguirao) [reply dali/kernels/reduce/reduce_setup_utils.h]: Done

**Comment by jantonguirao** [comment]
!build

**Pull request merged**